ContinousBenchmarking.md

```diff
@@ -0,0 +1,175 @@
+Continuous Benchmarking System
+======
+Kevin Cheng
+Initial Proposal 11/6/18
+
+
+# Background
+
+## Context
+Benchmarking an algorithm is a compute intensive and time consuming chore that is essential for guaranteeing performance. 
+An algorithm can unknowingly drift towards an undesirable state if the laborious task of benchmarking is neglected for
+too long.  
+
+## Problem Statement
+We want to create a continous benchmarking system to automate this task.  
+
+## Scope
+Problem starts with a code changes to Garage. The system needs to run benchmarking at least nightly, generate a results page, and deploy that page.  Problem ends with a pubically viewable results page.
+
+
+## Goals and Non-Goals
+*This version of the proposal will focus on a minimally viable system. These goals will evolve once the initial goal is accomplished and we decide to bring in more complexity.
+#### Goals:
+- Update benchmarking scripts so they can be started in a consistent format and run to completion without manual intervention.
+- Focus on getting system up and running on a single machine.
+- Design a simple system for deploying the results page.
+- Allow garage developers to detect regressions in algorithm performance caused by code changes
+
+#### Non-Goals: 
+- Worry about scalability too early.  MVP is top priority.
+- Design for frugality over simplicity
+- Manage web servers
+
+## Design Overview
+*The input and output interfaces of the benchmarking scripts must be normalized (start commands, and results).
+This initial design will use a cron job script ran on a single local machine. On a nightly basis the script shall:
+1. pull the latest garage code. 
+2. Run commands to launch benchmarking jobs
+3. Detect when the jobs are completed
+4. Run script to generate results page with plots.
+5. Deploy results.html to Github pages or similar.
+
+### Options:
+
+#### Python
+Perhaps we can just write a single Python script to accomplish steps 1 - 3? This way we can import the benchmarking scripts as modules, and run it like that.  
+
+#### Shell 
+We need to launch the conda env after pulling the latest garage.  Maybe this workflow would be more appropriate as a shell script.  
+
+For step 4, I am thinking we can just have a html/js template, and use grunt/gruntfile to do some simple replacements of data.
+
+For step 5, There seems to be many options, I'd appreciate all suggestions!
+
+
+
+## Detailed Design
+Rough outline of script:
+    
+    #pull latest garage -- something like this.
+    git clone https://github.com/rlworkgroup/garage.git 
```

Comment:
You probably want to use Docker for this. We've already covered all these steps in Dockerfile.ci. You will need to make a new version based on nvidia-docker if you want to make a Dockerfile which works on a GPU machine (e.g. for DQN with pixels)